File: dynamicArray.py
from collections import deque

class Graph:
    def __init__(self, size: int):
        """Initializes the graph with a given size (number of vertices)."""
        self.size = size
        self.adj_matrix = [[0] * size for _ in range(size)]
        self.vertex_data = [''] * size  

    def add_edge(self, u: int, v: int):
        """Adds an **undirected** edge between vertices `u` and `v`."""
        if 0 <= u < self.size and 0 <= v < self.size:
            self.adj_matrix[u][v] = 1
            self.adj_matrix[v][u] = 1

    def add_vertex_data(self, vertex: int, data: str):
        """Assigns data (labels) to a vertex."""
        if 0 <= vertex < self.size:
            self.vertex_data[vertex] = data

    def __str__(self):
        """Returns a string representation of the adjacency matrix."""
        matrix_str = "\n".join([" ".join(map(str, row)) for row in self.adj_matrix])
        vertex_str = "\n".join([f"Vertex {i}: {data}" for i, data in enumerate(self.vertex_data)])
        return f"Adjacency Matrix:\n{matrix_str}\n\nVertex Data:\n{vertex_str}"

    def is_cyclic_dfs(self, v: int, visited: list, parent: int) -> bool:
        """Helper DFS function for cycle detection."""
        visited[v] = True

        for neighbor in range(self.size):
            if self.adj_matrix[v][neighbor]:  # If there's an edge
                if not visited[neighbor]:
                    if self.is_cyclic_dfs(neighbor, visited, v):
                        return True
                elif neighbor != parent:  # Cycle found
                    return True

        return False

    def is_cyclic(self) -> bool:
        """Checks if the graph contains a cycle using DFS traversal."""
        visited = [False] * self.size

        for node in range(self.size):
            if not visited[node]:
                if self.is_cyclic_dfs(node, visited, -1):
                    return True
        return False


# Example Usage
g = Graph(7)

# Add vertex labels
for i, label in enumerate(['A', 'B', 'C', 'D', 'E', 'F', 'G']):
    g.add_vertex_data(i, label)

# Add edges
edges = [(3, 0), (0, 2), (0, 4), (4, 2), (2, 5), (2, 1), (2, 6), (1, 5)]
for u, v in edges:
    g.add_edge(u, v)

# Print graph
print(g)

# Check if graph contains a cycle
print("\nGraph has cycle:", g.is_cyclic())

File: dynamicArray.cpp
#include <iostream>
#include <vector>
#include <string>
#include <limits>

class Graph {
public:
    Graph(int size) {
        adj_matrix.resize(size, std::vector<int>(size, 0)); // Initialize adjacency matrix with 0
        vertex_data.resize(size, ""); // Vertex labels
        this->size = size;
    }

    void add_edge(int u, int v, int weight) {
        if (u >= 0 && u < size && v >= 0 && v < size) {
            adj_matrix[u][v] = weight; // Directed edge
        }
    }

    void add_vertex_data(int vertex, const std::string& data) {
        if (vertex >= 0 && vertex < size) {
            vertex_data[vertex] = data;
        }
    }

    std::tuple<bool, std::vector<int>, std::vector<int>> bellman_ford(const std::string& start_vertex_data) {
        int start_vertex = find_vertex_index(start_vertex_data);
        std::vector<int> distances(size, std::numeric_limits<int>::infinity()); // Initialize distances as infinity
        std::vector<int> predecessors(size, -1); // Predecessors to track paths
        distances[start_vertex] = 0; // Distance to start vertex is 0

        // Relax edges |V| - 1 times
        for (int i = 0; i < size - 1; ++i) {
            bool updated = false;
            for (int u = 0; u < size; ++u) {
                for (int v = 0; v < size; ++v) {
                    if (adj_matrix[u][v] != 0) { // If there's an edge from u to v
                        int new_dist = distances[u] + adj_matrix[u][v];
                        if (new_dist < distances[v]) { // Relax the edge
                            distances[v] = new_dist;
                            predecessors[v] = u;
                            updated = true;
                        }
                    }
                }
            }
            if (!updated) {
                break; // Stop early if no updates in this pass
            }
        }

        // Negative cycle detection (after all relaxations)
        for (int u = 0; u < size; ++u) {
            for (int v = 0; v < size; ++v) {
                if (adj_matrix[u][v] != 0 && distances[u] + adj_matrix[u][v] < distances[v]) {
                    return {true, {}, {}}; // Negative cycle detected
                }
            }
        }

        return {false, distances, predecessors};
    }

    std::string get_path(const std::vector<int>& predecessors, const std::string& start_vertex_data, const std::string& end_vertex_data) {
        int start_vertex = find_vertex_index(start_vertex_data);
        int end_vertex = find_vertex_index(end_vertex_data);

        std::vector<std::string> path;
        int current = end_vertex;
        while (current != -1) {
            path.insert(path.begin(), vertex_data[current]); // Insert at the beginning of the list
            current = predecessors[current];
            if (current == start_vertex) { // Stop when we reach the start vertex
                path.insert(path.begin(), vertex_data[start_vertex]);
                break;
            }
        }

        std::string path_str;
        for (const auto& vertex : path) {
            path_str += vertex + "->";
        }

        if (!path_str.empty()) {
            path_str.pop_back(); // Remove last '->'
        }

        return path_str;
    }

private:
    std::vector<std::vector<int>> adj_matrix; // Adjacency matrix
    std::vector<std::string> vertex_data;     // Vertex labels
    int size;

    int find_vertex_index(const std::string& vertex_data_str) {
        for (int i = 0; i < size; ++i) {
            if (vertex_data[i] == vertex_data_str) {
                return i;
            }
        }
        return -1; // Shouldn't happen if the vertex exists
    }
};

int main() {
    Graph g(5);

    // Add vertices with labels
    g.add_vertex_data(0, "A");
    g.add_vertex_data(1, "B");
    g.add_vertex_data(2, "C");
    g.add_vertex_data(3, "D");
    g.add_vertex_data(4, "E");

    // Add directed edges with weights
    g.add_edge(3, 0, 4);  // D -> A, weight 4
    g.add_edge(3, 2, 7);  // D -> C, weight 7
    g.add_edge(3, 4, 3);  // D -> E, weight 3
    g.add_edge(0, 2, 4);  // A -> C, weight 4
    g.add_edge(2, 0, -3); // C -> A, weight -3
    g.add_edge(0, 4, 5);  // A -> E, weight 5
    g.add_edge(4, 2, 3);  // E -> C, weight 3
    g.add_edge(1, 2, -4); // B -> C, weight -4
    g.add_edge(4, 1, 2);  // E -> B, weight 2

    // Running the Bellman-Ford algorithm from D to all vertices
    std::cout << "\nThe Bellman-Ford Algorithm starting from vertex D:" << std::endl;
    auto [negative_cycle, distances, predecessors] = g.bellman_ford("D");

    if (!negative_cycle) {
        for (int i = 0; i < distances.size(); ++i) {
            if (distances[i] != std::numeric_limits<int>::infinity()) {
                std::string path = g.get_path(predecessors, "D", g.vertex_data[i]);
                std::cout << "Shortest path from D to " << g.vertex_data[i] << ": " << path << ", Distance: " << distances[i] << std::endl;
            } else {
                std::cout << "No path from D to " << g.vertex_data[i] << ", Distance: Infinity" << std::endl;
            }
        }
    } else {
        std::cout << "Negative weight cycle detected. Cannot compute shortest paths." << std::endl;
    }

    return 0;
}
